File: 08FlowControl/match/binding.cs
using System;
public class Program
{
    public static uint Age() {
        return 15u;
    }

    public static int? SomeNumber(){
        return 42;
    }
    //guard,就是match 然后用if的方式。
    public static void Main(string[] args)
    {
        Console.WriteLine("This is C# Language");
        var age = Age();
        var message = age switch {
            0 => "I haven't celebrated my first birthday yet",
            >= 1 and <= 12 => $"I'm a child of age {age}",
            >= 13 and <= 19 => $"I'm a teen of age {age}",
            _ => $"I'm an old person of age {age}"
        };
        Console.WriteLine(message);

        int? number = SomeNumber();
        message = number switch{
            42=>$"The Answer: {number}!",
            int n=>$"Not interesting... {n}",
            _=>"",
        };
        if( message != "") Console.WriteLine(message);
    }
}

File: 08FlowControl/match/binding.rs
// `age` 函数，返回一个 `u32` 值。
fn age() -> u32 {
    15
}

fn some_number() -> Option<u32> {
	Some(42)
}
fn main() {
    println!("Tell me what type of person you are");

    match age() { //针对卫语句，这里用if 也行 不过写起来麻烦一点。
        0             => println!("I haven't celebrated my first birthday yet"),
        // 可以直接匹配（`match`） 1 ..= 12，但那样的话孩子会是几岁？
        // 相反，在 1 ..= 12 分支中绑定匹配值到 `n` 。现在年龄就可以读取了。
        n @ 1  ..= 12 => println!("I'm a child of age {:?}", n),//缺乏in的判断语句
        n @ 13 ..= 19 => println!("I'm a teen of age {:?}", n),
        // 不符合上面的范围。返回结果。
        n             => println!("I'm an old person of age {:?}", n),
    }
    
    match some_number() {
        // 得到 `Some` 可变类型，如果它的值（绑定到 `n` 上）等于 42，则匹配。
        Some(n @ 42) => println!("The Answer: {}!", n),
        // 匹配任意其他数字。
        Some(n)      => println!("Not interesting... {}", n),
        // 匹配任意其他值（`None` 可变类型）。
        _            => (),
    }
    match some_number() {//Option自带一个None，必须处理的类型。
        // 得到 `Some` 可变类型，如果它的值（绑定到 `n` 上）等于 42，则匹配。
        Some(n) if n==42 => println!("The Answer: {}!", n),
        // 匹配任意其他数字。
        Some(n)      => println!("Not interesting... {}", n),
        // 匹配任意其他值（`None` 可变类型）。
        _            => (),
    }
}
